zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/ZkClient.java

```diff
@@ -257,12 +274,22 @@ public void subscribeDataChanges(String path, IZkDataListener listener) {
         }
       }
     }
-    watchForData(path);
+
```

Comment:
I don't think we can switch this order here. Immediately after the watcher is installed, the event might come. If it happens before _dataListener put is done, then we miss that event.

Reply:
Agreed. Maybe we could do this: add listeners first. If watch fails, we remove listeners?

Reply:
@jiajunwang , this is a very good point. 
The how about what @pkuwm suggested, add listener first, if install watch fails, remove this listener?

Reply:
adopt the approach proposed by @pkuwm.